src/index.js

```diff
@@ -1,41 +1,194 @@
 const spdy = require('libp2p-spdy')
 const multiaddr = require('multiaddr')
+const mafmt = require('mafmt')
 const EE = require('events').EventEmitter
 
 exports = module.exports
 
-exports.Node = function Node (peerInfo) {
+const OFFLINE_ERROR = new Error('The libp2p node is not started yet')
+
+exports.Node = function Node (pInfo, pBook) {
   if (!(this instanceof Node)) {
-    return new Node(peerInfo)
+    return new Node(pInfo, pBook)
   }
-  if (!peerInfo) {
-    peerInfo = new Peer()
-    peerInfo.multiaddr.add(multiaddr('/ip4/0.0.0.0/tcp/0'))
+
+  if (!pInfo) {
+    pInfo = new PeerInfo()
+    pInfo.multiaddr.add(multiaddr('/ip4/0.0.0.0/tcp/0'))
   }
 
+  if (!pBook) {
+    pBook = new PeerBook()
+  }
+
+  this.peerInfo = pInfo
+  this.peerBook = pBook
+
   // Swarm
-  this.swarm = new Swarm(peerInfo)
-  this.swarm.transport.add('ws', new WS())
-  this.swarm.transport.add('tcp', new TCP())
+  this.swarm = new Swarm(pInfo)
   this.swarm.connection.addStreamMuxer(spdy)
   this.swarm.connection.reuse()
 
-  this.start = (done) => {
-    this.swarm.listen(done)
+  this.swarm.on('peer-mux-established', (peerInfo) => {
+    this.peerBook.put(peerInfo)
+  })
+
+  this.swarm.on('peer-mux-closed', (peerInfo) => {
+    this.peerBook.removeByB58String(peerInfo.id.toB58String())
+  })
+
+  this.isOnline = false
+
+  this.start = (callback) => {
+    const ws = new WS()
+    const tcp = new TCP()
+
+    // Do not activate the dialer if no listener is going to be present
+    if (ws.filter(this.peerInfo.multiaddrs).length > 0) {
+      this.swarm.transport.add('ws', new WS())
+    }
+    if (tcp.filter(this.peerInfo.multiaddrs).length > 0) {
+      this.swarm.transport.add('tcp', new TCP())
+    }
+
+    this.swarm.listen((err) => {
+      if (err) {
+        return callback(err)
+      }
+      this.isOnline = true
+      callback()
+    })
+  }
+
+  this.stop = (callback) => {
+    this.isOnline = false
+    this.swarm.close(callback)
+  }
+
+  this.dialById = (id, protocol, callback) => {
+    if (typeof protocol === 'function') {
+      callback = protocol
+      protocol = undefined
+    }
+
+    if (!this.isOnline) {
+      return callback(OFFLINE_ERROR)
+    }
+    // NOTE, these dialById only works if a previous dial
+    // was made until we have PeerRouting
+    // TODO support PeerRouting when it is Ready
+    callback(new Error('not implemented yet'))
+  }
+
+  this.dialByMultiaddr = (maddr, protocol, callback) => {
+    if (typeof protocol === 'function') {
+      callback = protocol
+      protocol = undefined
+    }
+
+    if (!this.isOnline) {
+      return callback(OFFLINE_ERROR)
+    }
+
+    if (typeof maddr === 'string') {
+      maddr = multiaddr(maddr)
+    }
+
+    if (!mafmt.IPFS.matches(maddr.toString())) {
+      return callback(new Error('multiaddr not valid'))
+    }
+
+    let ipfsIdB58String
+    maddr.stringTuples().forEach((tuple) => {
+      if (tuple[0] === 421) {
+        ipfsIdB58String = tuple[1]
+      }
+    })
```

Comment:
This could be written easier using either native `.filter` or [`lodash.find`]:

```js
const ipfsId = _.find(maddr.stringTuples(), (tuple) => tuple[0] === 421)
const ipfsIdB58String = ipfsId ? ipfsId[1] : null
```